File: src/16/day16.py
from collections import namedtuple
from typing import Generator

# index: int, direction: int, split: bool
StepResult = namedtuple('StepResult', ['index', 'direction', 'split'])

# direction is one of: 1, -1, x_bound, -x_bound
def step(index: int, grid: list[str], x_bound: int, direction: int) -> StepResult:
    travelling_x = abs(direction) == 1
    grid_bound = 0 if direction < 0 else index + x_bound if travelling_x else len(grid)

    if direction < 0:
        if travelling_x:
            grid_bound = (index // x_bound) * x_bound - 1
        else:
            grid_bound = -1
    else:
        if travelling_x:
            grid_bound = (index // x_bound + 1) * x_bound
        else:
            grid_bound = len(grid)

    i = index
    for i in range(index + direction, grid_bound, direction):
        match grid[i]:
            case '|' if travelling_x:
                return StepResult(i, x_bound, True)
            case '-' if not travelling_x:
                return StepResult(i, 1, True)
            case '/':
                m = x_bound if travelling_x else 1
                s = -1 if direction > 0 else 1
                return StepResult(i, s * m, False)
            case '\\':
                m = x_bound if travelling_x else 1
                s = -1 if direction < 0 else 1
                return StepResult(i, s * m, False)
    return StepResult(i, 0, False)

def energy(start_index: int, starting_direction: int, grid: list[str], x_bound: int) -> int:
    traversed_points = set()
    traversed_directions = {(start_index, starting_direction)}
    paths = {(start_index, starting_direction)}
    while paths:
        new_paths = set()
        for index, direction in paths:
            step_result = step(index, grid, x_bound, direction)
            if step_result.direction == 0:
                pass
            elif step_result.split:
                if (step_result.index, step_result.direction) not in traversed_directions:
                    new_paths.add((step_result.index, step_result.direction))
                    traversed_directions.add((step_result.index, step_result.direction))
                if (step_result.index, -step_result.direction) not in traversed_directions:
                    new_paths.add((step_result.index, -step_result.direction))
                    traversed_directions.add((step_result.index, -step_result.direction))
            else:
                if (step_result.index, step_result.direction) not in traversed_directions:
                    new_paths.add((step_result.index, step_result.direction))
            traversed_points |= set(range(index, step_result.index + direction, direction))
        paths = new_paths
    return len(traversed_points)

def part1(filename: str) -> int:
    with open(filename, 'rt') as f:
        lines = [x.strip() for x in f.readlines()]
    x_bound = len(lines[0])
    grid = sum([list(x) for x in lines], [])
    match grid[0]:
        case '.' | '-':
            starting_direction = 1
        case '|' | '\\':
            starting_direction = x_bound
        case '/':
            starting_direction = -x_bound
    return energy(0, starting_direction, grid, x_bound)

def all_directions(index: int, x_bound: int, y_bound: int) -> Generator[int, None, None]:
    if index % x_bound == 0:
        yield 1
    elif index % x_bound == x_bound - 1:
        yield -1
    if index // x_bound == 0:
        yield x_bound
    elif index // x_bound == y_bound - 1:
        yield -x_bound

def part2(filename: str) -> int:
    with open(filename, 'rt') as f:
        lines = [x.strip() for x in f.readlines()]
    x_bound = len(lines[0])
    y_bound = len(lines)
    grid = sum([list(x) for x in lines], [])
    edges = (set(range(0, x_bound)) |
        set(range(0, len(grid), x_bound)) |
        set(range(x_bound - 1, len(grid), x_bound)) |
        set(range(len(grid) - x_bound, len(grid))))
    return max(
        energy(edge, direction, grid, x_bound)
        for edge in edges
        for direction in all_directions(edge, x_bound, y_bound)
    )

if __name__ == '__main__':
    print(part1('day16_input.txt'))
    print(part2('day16_input.txt'))

File: src/16/day16.rs
use std::collections::BTreeSet;

#[derive(Clone, Copy, Debug)]
pub enum Mirror {
    Empty,
    Vertical,
    Horizontal,
    UpRight,
    UpLeft
}

#[derive(Clone, Copy, Debug, PartialOrd, Ord, PartialEq, Eq)]
pub enum Direction {
    Right,
    Left,
    Down,
    Up
}

type StepResult = (usize, Option<Direction>, bool);

fn point_range(start: i64, stop: i64, step: i64) -> Vec<usize> {
    let mut points: Vec<usize> = Vec::new();
    let mut index = start;
    if step < 0 {
        while index > stop {
            points.push(index as usize);
            index += step;
        }
    } else {
        while index < stop {
            points.push(index as usize);
            index += step;
        }
    }
    return points;
}

fn direction_offset(direction: Direction, x_bound: usize) -> i64 {
    match direction {
        Direction::Left  => -1,
        Direction::Up    => -(x_bound as i64),
        Direction::Right => 1,
        Direction::Down  => x_bound as i64
    }
}

fn reverse(direction: Direction) -> Direction {
    match direction {
        Direction::Left  => Direction::Right,
        Direction::Up    => Direction::Down,
        Direction::Right => Direction::Left,
        Direction::Down  => Direction::Up
    }
}

fn against_up_right(direction: Direction) -> Direction {
    match direction {
        Direction::Left  => Direction::Down,
        Direction::Up    => Direction::Right,
        Direction::Right => Direction::Up,
        Direction::Down  => Direction::Left
    }
}

fn against_up_left(direction: Direction) -> Direction {
    match direction {
        Direction::Left  => Direction::Up,
        Direction::Up    => Direction::Left,
        Direction::Right => Direction::Down,
        Direction::Down  => Direction::Right
    }
}

fn step(index: usize, grid: &Vec<Mirror>, x_bound: usize, direction: Direction) -> StepResult {
    let travelling_x = match direction {
        Direction::Right | Direction::Left => true,
        _ => false
    };
    let grid_bound = match direction {
        Direction::Left  => ((index / x_bound) * x_bound) as i64 - 1,
        Direction::Up    => -1,
        Direction::Right => ((index / x_bound + 1) * x_bound) as i64,
        Direction::Down  => grid.len() as i64
    };
    let increment = direction_offset(direction, x_bound);
    let mut j = index as i64;
    for i in point_range(j + increment, grid_bound, increment) {
        j = i as i64;
        match grid[i as usize] {
            Mirror::Vertical if travelling_x => return (i, Some(Direction::Down), true),
            Mirror::Horizontal if !travelling_x => return (i, Some(Direction::Right), true),
            Mirror::UpRight => return (i, Some(against_up_right(direction)), false),
            Mirror::UpLeft => return (i, Some(against_up_left(direction)), false),
            _ => {}
        }
    }
    return (j as usize, None, false);
}

fn energy(start_index: usize, starting_direction: Direction, grid: &Vec<Mirror>, x_bound: usize) -> usize {
    let mut traversed_points = BTreeSet::<usize>::new();
    let mut traversed_directions = BTreeSet::<(usize, Direction)>::new();
    let mut paths = BTreeSet::<(usize, Direction)>::new();
    paths.insert((start_index, starting_direction));
    while paths.len() > 0 {
        let mut new_paths = BTreeSet::<(usize, Direction)>::new();
        for &(index, direction) in &paths {
            let step_result = step(index, &grid, x_bound, direction);
            match step_result.1 {
                Some(d) => {
                    if step_result.2 {
                        if !traversed_directions.contains(&(step_result.0, d)) {
                            new_paths.insert((step_result.0, d));
                            traversed_directions.insert((step_result.0, d));
                        }
                        if !traversed_directions.contains(&(step_result.0, reverse(d))) {
                            new_paths.insert((step_result.0, reverse(d)));
                            traversed_directions.insert((step_result.0, reverse(d)));
                        }
                    } else {
                        if !traversed_directions.contains(&(step_result.0, d)) {
                            new_paths.insert((step_result.0, d));
                        }
                    }
                },
                None => {}
            }
            for i in point_range(
                index as i64,
                step_result.0 as i64 + direction_offset(direction, x_bound),
                direction_offset(direction, x_bound))
            {
                traversed_points.insert(i);
            }
        }
        paths = new_paths;
    }
    return traversed_points.len();
}

fn part1(filename: &str) -> usize {
    let file_lines = std::fs::read_to_string(filename).expect("File not found");
    let lines = file_lines.trim().split('\n').map(|x| x.trim()).collect::<Vec<_>>();
    let x_bound: usize = lines[0].len();
    let mut grid: Vec<Mirror> = Vec::new();
    for line in lines {
        grid.extend(line.chars().map(
            |tile|
            match tile {
                '|' => Mirror::Vertical,
                '-' => Mirror::Horizontal,
                '/' => Mirror::UpRight,
                '\\' => Mirror::UpLeft,
                _ => Mirror::Empty
            }
        ));
    }
    let starting_direction = match grid[0] {
        Mirror::Empty | Mirror::Horizontal => Direction::Right,
        Mirror::Vertical | Mirror::UpLeft => Direction::Down,
        Mirror::UpRight => Direction::Up
    };
    return energy(0, starting_direction, &grid, x_bound);
}

fn all_directions(index: usize, x_bound: usize, y_bound: usize, mirror: Mirror) -> Vec<Direction> {
    let mut directions: Vec<Direction> = Vec::new();
    if index % x_bound == 0 {
        match mirror {
            Mirror::Empty | Mirror::Horizontal => directions.push(Direction::Right),
            Mirror::Vertical => {
                directions.push(Direction::Up);
                directions.push(Direction::Down);
            },
            Mirror::UpRight => directions.push(Direction::Up),
            Mirror::UpLeft => directions.push(Direction::Down)
        }
    } else if index % x_bound == x_bound - 1 {
        match mirror {
            Mirror::Empty | Mirror::Horizontal => directions.push(Direction::Left),
            Mirror::Vertical => {
                directions.push(Direction::Up);
                directions.push(Direction::Down);
            },
            Mirror::UpRight => directions.push(Direction::Down),
            Mirror::UpLeft => directions.push(Direction::Up)
        }
    }
    if index / x_bound == 0 {
        match mirror {
            Mirror::Empty | Mirror::Horizontal => directions.push(Direction::Down),
            Mirror::Vertical => {
                directions.push(Direction::Left);
                directions.push(Direction::Right);
            },
            Mirror::UpRight => directions.push(Direction::Left),
            Mirror::UpLeft => directions.push(Direction::Right)
        }
    } else if index / x_bound == y_bound - 1 {
        match mirror {
            Mirror::Empty | Mirror::Horizontal => directions.push(Direction::Up),
            Mirror::Vertical => {
                directions.push(Direction::Left);
                directions.push(Direction::Right)
            },
            Mirror::UpRight => directions.push(Direction::Right),
            Mirror::UpLeft => directions.push(Direction::Left)
        }
    }
    return directions;
}

fn part2(filename: &str) -> usize {
    let file_lines = std::fs::read_to_string(filename).expect("File not found");
    let lines = file_lines.trim().split('\n').map(|x| x.trim()).collect::<Vec<_>>();
    let x_bound: usize = lines[0].len();
    let y_bound: usize = lines.len();
    let mut grid: Vec<Mirror> = Vec::new();
    for line in lines {
        grid.extend(line.chars().map(
            |tile|
            match tile {
                '|' => Mirror::Vertical,
                '-' => Mirror::Horizontal,
                '/' => Mirror::UpRight,
                '\\' => Mirror::UpLeft,
                _ => Mirror::Empty
            }
        ));
    }
    let edges: BTreeSet<usize> = point_range(0, x_bound as i64, 1).into_iter()
        .chain(point_range(0, grid.len() as i64, x_bound as i64).into_iter())
        .chain(point_range(x_bound as i64 - 1, grid.len() as i64, x_bound as i64).into_iter())
        .chain(point_range((grid.len() - x_bound) as i64, grid.len() as i64, 1).into_iter())
        .collect();
    return edges
        .iter()
        .map(
            |&edge|
            all_directions(edge, x_bound, y_bound, grid[edge])
            .iter()
            .map(
                |&direction|
                energy(edge, direction, &grid, x_bound)
            )
            .fold(0, |acc, x| if acc > x { acc } else { x })
        )
        .fold(0, |acc, x| if acc > x { acc } else { x });
}

fn main() {
    println!("{}", part1("day16_input.txt"));
    println!("{}", part2("day16_input.txt"));
}
